doc/ps_design.md

```diff
@@ -0,0 +1,375 @@
+
+**Question 3** async-SGD and sync-SGD
+
+
+There are two ways to support async-SGD:
+ 
+- Calling `apply_gradient` inside `push_gradient` gRPC service. The gradients will not be put into the queue.
+- Putting gradients into the queue, and optimizer gets gradients from the queue immediately to `apply_gradient`.
+
+In the first way, there may be several gRPC threads running in parallel. This will introduce race condition on parameter updating, some gradients may be overwrited.
+
+The second way ensure each gradient could be applied, and decoupling these two procedures, `push_gradient` of worker and `apply_gradient` of optimizer. But the second way introduces more staleness in updating model, and may influence the final training accuracy.
+
+We may consider the second way later.
+
+In sync-SGD, optimizer needs to wait for a certain number of gradients, and then get the gradient after addition. We could implement a customized queue structure to support such logic efficiently.
+
+### Optimizer
+
+Optimizer gets a gradient from the gradient queue, and query to get the corresponding parameter from KVStore. Then it will apply the gradient to the parameter. It has a `tf.keras.optimizer` instance inside.
+
+Many `tf.keras.optimizer` subclasses, such as `Adam` and `Adagrad` allocate and manage additional variables associated with the variables to train.  These are called `Slots`.
+
+Embedding table slots are stored at KVStore, and other common parameter slots are stored and managed by `tf.keras.optimizer`.
+
+The embedding table slot is also a embedding table data structure. For example, a embedding table parameter with name `embedding_layer0`, we will create a corresponding `embedding_layer0-momentum` EmbeddingTable object in `KVStore.embedding_table_db`.
+
+
+### RPC Service
+
+PServer provide RPC service for workers.
+
+Since pserver will store a subset of the full model. And a worker will push/pull a submodel from the pserver. 
+
+However, embedding table is initialized lazily in pserver, worker should also send embedding table information to pserver. We have to add another filed to describe embedding table related information.
+
+The model message is defined as following:
+
+```proto
+
+message EmbeddingTableInfo{
+    string name = 1;
+    repeated int64 dims = 2;
+    string initializer = 3;
+}
+
+message Model {
+    int64 version = 1;
+    repeated Tensor tensors = 2;
+    repeated EmbeddingTableInfo embedding_table_info = 3;
+}
+```
+
+Model could also be used as gradients collection.
+
+
+
+So the RPC service will be defined as following:
+
+```proto
+
+service PServer{
+    rpc push_model(Model) returns (google.protobuf.Empty) {}
+    rpc pull_model(Model) returns (Model) {}
+    rpc pull_embedding_vector(Tensor) returns (EmbeddingResponse) {}
+    rpc push_gradient(Model) returns (google.protobuf.Empty)
```

Comment:
Question 3, name convention

- `compute_task` following python style
- or `ComputeTask`

If we use gRPC only with Python, personally, I think it's better to follow Python pep8 style..
There is also another discussion [here](https://github.com/grpc/grpc/issues/5014)